### ISSUE TITLE
[Question] Is it possible to get a window to run in windows forms?

### ISSUE THREAD

**Issue by levonravel** [mellinoe/veldrid#155]
I am currently looking into Veldrid to run in my project. Its more of a learning experience then anything I am currently using OpenTK and they have something called  GLControl basically just pops out as a component. Here is a [screenshot](https://snag.gy/RumH0p.jpg) of what I am talking about. Thank you for your time on looking over this question -Levon

**Comment by zaafar** [comment]
Why would you want to use windows form rather than Sdl2Window?

**Comment by ppc2017** [comment]
> 
> 
> Out of curiosity, why would you want to use windows form rather than Sdl2Window?

I don't know about the OP, but I would prefer windows forms because many basic things that are necessary for desktop applications are missing from Veldrid SDL2Window. For example: custom window icon, custom cursors, drag and drop operations etc...

**Comment by mellinoe** [comment]
Yes, it's simple to get Veldrid running in a WinForms app, or a number of other UI frameworks. Nowadays, you might be better off using WPF since it's more powerful and a bit more modern.

To integrate Veldrid, you just need to create a Swapchain that targets the Win32 handle (`Control.Handle`) of your form. Here is a really basic snippet of code that creates a GraphicsDevice and Swapchain, and then repeatedly clear the window to blue.

```C#
public partial class Form1 : Form
{
    private GraphicsDevice _gd;
    private Swapchain _sc;
    private CommandList _cl;

    public Form1()
    {
        InitializeComponent();
    }

    private void Form1_Load(object sender, EventArgs e)
    {
        _gd = GraphicsDevice.CreateD3D11(new GraphicsDeviceOptions());
        SwapchainSource source = SwapchainSource.CreateWin32(Handle, Marshal.GetHINSTANCE(typeof(Form1).Module));
        _sc = _gd.ResourceFactory.CreateSwapchain(
            new SwapchainDescription(source, (uint)Width, (uint)Height, null, false));
        _cl = _gd.ResourceFactory.CreateCommandList();
    }

    protected override void OnPaint(PaintEventArgs e)
    {
        _cl.Begin();
        _cl.SetFramebuffer(_sc.Framebuffer);
        _cl.ClearColorTarget(0, RgbaFloat.CornflowerBlue);
        _cl.End();
        _gd.SubmitCommands(_cl);
        _gd.SwapBuffers(_sc);
    }
}
```

As you can see, the `Swapchain` object itself is created by itself from the control's handle separate from the GraphicsDevice. It's possible to create multiple Swapchains that each target different panels in your application if you want (for separate 3D views, etc.).

Let me know if that helps, or if there is anything else I can clarify.

**Comment by levonravel** [comment]
Honestly I had already threw myself into trying to learn so I started with what I know. Just so happens that windows forms was on the list :)

**Comment by levonravel** [comment]
> Out of curiosity, why would you want to use windows form rather than Sdl2Window?

I just happened to started with windows forms as its something I know. -Levon

**Comment by levonravel** [comment]
> > Out of curiosity, why would you want to use windows form rather than Sdl2Window?
> 
> I don't know about the OP, but I would prefer windows forms because many basic things that are necessary for desktop applications are missing from Veldrid SDL2Window. For example: custom window icon, custom cursors, drag and drop operations etc...

Very good point, but was not the main reason haha.

**Comment by levonravel** [comment]
@mellinoe, Thank you very much. I will start intergration now as I was using OpenTK and GL4
